File: ts/src/refs.ts
/** One `$` form found in an authored string (SPEC §6). */
export interface Reference {
  name: string;
  hasDefault: boolean;
}

const NAME_START = /[A-Za-z_]/;
const NAME_CHAR = /[A-Za-z0-9_]/;

/**
 * The variable references of an authored string in order of appearance. Used
 * for provenance of a value that already loaded, so malformed forms are simply
 * skipped rather than reported.
 */
export function scanReferences(s: string): Reference[] {
  const out: Reference[] = [];
  let i = 0;
  while (i < s.length) {
    if (s[i] !== "$" || i + 1 >= s.length) {
      i++;
      continue;
    }
    const next = s[i + 1];
    if (next === "$") {
      i += 2;
    } else if (next === "{") {
      const end = s.indexOf("}", i + 2);
      if (end === -1) return out;
      const inner = s.slice(i + 2, end);
      const colon = inner.indexOf(":");
      out.push(
        colon === -1
          ? { name: inner, hasDefault: false }
          : { name: inner.slice(0, colon), hasDefault: true },
      );
      i = end + 1;
    } else if (NAME_START.test(next)) {
      let j = i + 1;
      while (j < s.length && NAME_CHAR.test(s[j])) j++;
      out.push({ name: s.slice(i + 1, j), hasDefault: false });
      i = j;
    } else {
      i++;
    }
  }
  return out;
}

File: go/refs.go
package entryconf

// reference is one $ form found in an authored string (SPEC §6).
type reference struct {
	name       string
	hasDefault bool
}

// scanReferences lists the variable references of an authored string in order
// of appearance, tolerating malformed forms (it is used for provenance of a
// value that already loaded, so the string is known to be well-formed).
func scanReferences(s string) []reference {
	var out []reference
	i := 0
	for i < len(s) {
		if s[i] != '$' || i+1 >= len(s) {
			i++
			continue
		}
		switch next := s[i+1]; {
		case next == '$':
			i += 2
		case next == '{':
			end := indexByteFrom(s, '}', i+2)
			if end < 0 {
				return out
			}
			inner := s[i+2 : end]
			name, hasDefault := inner, false
			if colon := indexByteFrom(inner, ':', 0); colon >= 0 {
				name, hasDefault = inner[:colon], true
			}
			out = append(out, reference{name: name, hasDefault: hasDefault})
			i = end + 1
		case isNameStart(next):
			j := i + 1
			for j < len(s) && isNameChar(s[j]) {
				j++
			}
			out = append(out, reference{name: s[i+1 : j]})
			i = j
		default:
			i++
		}
	}
	return out
}

func indexByteFrom(s string, c byte, from int) int {
	for i := from; i < len(s); i++ {
		if s[i] == c {
			return i
		}
	}
	return -1
}
